File: lib/scoring.py
"""Score calculation and tiebreaker logic for the Masters Pool.

Rules summary (from Rules.md):
- Team score = sum of 5 golfers' scores to par.
- Missed cut: golfer's score freezes at their 36-hole total (still counts).
- DQ or WD (after tournament starts): entire team is disqualified.
- Tiebreakers:
  1. Lower score from the tier-4 (51+) golfer.
  2. Lower combined score from the two tier-2 (11-30) golfers.
  3. If still tied: split prize money.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from lib.models import GolferScore, GolferStatus, Team, TeamGolferSlot, TeamStatus, Tier


@dataclass
class ScoredTeam:
    """A team with its computed score and tiebreaker values."""

    team: Team
    total_score: Optional[int]  # None if disqualified
    is_disqualified: bool
    tier4_score: Optional[int]  # tiebreaker 1
    tier2_combined_score: Optional[int]  # tiebreaker 2
    rank: Optional[int] = None  # assigned after sorting


def calculate_team_score(
    team: Team,
    golfer_scores: dict[str, GolferScore],
) -> ScoredTeam:
    """Calculate a team's total score and check for disqualification.

    Args:
        team: The team to score.
        golfer_scores: Map of espn_id -> GolferScore for all golfers in the tournament.

    Returns:
        ScoredTeam with computed total and tiebreaker values.
    """
    total = 0
    tier4_score: int | None = None
    tier2_scores: list[int] = []

    for slot in team.golfers:
        golfer = golfer_scores.get(slot.golfer_id)

        if golfer is None:
            continue

        # DQ or WD after starting -> team is disqualified
        if golfer.is_eliminated:
            return ScoredTeam(
                team=team,
                total_score=None,
                is_disqualified=True,
                tier4_score=None,
                tier2_combined_score=None,
            )

        # Use score_to_par (works for both active players and those who missed the cut,
        # since ESPN keeps their score frozen at their 36-hole total)
        score = golfer.score_to_par if golfer.score_to_par is not None else 0
        total += score

        # Track tiebreaker values
        if slot.tier == Tier.TIER_4:
            tier4_score = score
        elif slot.tier == Tier.TIER_2:
            tier2_scores.append(score)

    tier2_combined = sum(tier2_scores) if tier2_scores else None

    return ScoredTeam(
        team=team,
        total_score=total,
        is_disqualified=False,
        tier4_score=tier4_score,
        tier2_combined_score=tier2_combined,
    )


def rank_teams(
    teams: list[Team],
    golfer_scores: dict[str, GolferScore],
) -> list[ScoredTeam]:
    """Score all teams and return them ranked lowest-score-first.

    Disqualified teams sort to the bottom.

    Tiebreaker order:
      1. Lower total score wins.
      2. Lower tier-4 (51+) golfer score.
      3. Lower combined tier-2 (11-30) golfer scores.
      4. If still tied, they share the position (split prize money).
    """
    scored = [calculate_team_score(t, golfer_scores) for t in teams]

    def sort_key(st: ScoredTeam) -> tuple:
        if st.is_disqualified:
            return (1, 0, 0, 0)
        return (
            0,
            st.total_score if st.total_score is not None else 999,
            st.tier4_score if st.tier4_score is not None else 999,
            st.tier2_combined_score if st.tier2_combined_score is not None else 999,
        )

    scored.sort(key=sort_key)

    # Assign ranks (ties get the same rank; DQ'd teams are unranked)
    active_position = 0  # counts only non-DQ teams seen so far
    for st in scored:
        if st.is_disqualified:
            st.rank = None
            continue

        active_position += 1
        if active_position == 1:
            st.rank = 1
        else:
            # Find the previous non-DQ team
            prev = next(
                (s for s in reversed(scored[:scored.index(st)]) if not s.is_disqualified),
                None,
            )
            if prev and _is_tied(prev, st):
                st.rank = prev.rank
            else:
                st.rank = active_position

    return scored


def _is_tied(a: ScoredTeam, b: ScoredTeam) -> bool:
    """Check if two scored teams are tied after all tiebreakers."""
    return (
        a.total_score == b.total_score
        and a.tier4_score == b.tier4_score
        and a.tier2_combined_score == b.tier2_combined_score
    )


def check_team_disqualification(
    team: Team,
    golfer_scores: dict[str, GolferScore],
) -> bool:
    """Return True if any golfer on the team has WD or DQ status."""
    for slot in team.golfers:
        golfer = golfer_scores.get(slot.golfer_id)
        if golfer and golfer.is_eliminated:
            return True
    return False

File: src/lib/scoring.ts
/**
 * Score calculation and tiebreaker logic for the Masters Pool.
 *
 * Rules summary:
 * - Team score = sum of 5 golfers' scores to par.
 * - Missed cut: golfer's score freezes at their 36-hole total (still counts).
 * - DQ or WD (after tournament starts): entire team is disqualified.
 * - Tiebreakers:
 *   1. Lower score from the tier-4 (51+) golfer.
 *   2. Lower combined score from the two tier-2 (11-30) golfers.
 *   3. If still tied: split prize money.
 */

import type { GolferStatus, Tier } from "@/types";
import { DQ_STATUSES } from "./schema";

// Golfer score data needed for team scoring
export interface GolferScoreData {
  id: string;
  score_to_par: number | null;
  status: GolferStatus;
}

// Team golfer slot with tier assignment
export interface TeamGolferSlot {
  golfer_id: string;
  tier: Tier;
}

// Team with golfer slots
export interface TeamForScoring {
  id: string;
  golfers: TeamGolferSlot[];
}

// Result of scoring a team
export interface ScoredTeam {
  team: TeamForScoring;
  totalScore: number | null; // null if disqualified
  isDisqualified: boolean;
  tier4Score: number | null; // tiebreaker 1
  tier2CombinedScore: number | null; // tiebreaker 2
  rank: number | null; // assigned after sorting, null for DQ'd teams
}

/**
 * Check if a golfer status causes team disqualification.
 */
export function isEliminatedStatus(status: GolferStatus): boolean {
  return DQ_STATUSES.has(status);
}

/**
 * Calculate a team's total score and check for disqualification.
 */
export function calculateTeamScore(
  team: TeamForScoring,
  golferScores: Map<string, GolferScoreData>
): ScoredTeam {
  let total = 0;
  let tier4Score: number | null = null;
  const tier2Scores: number[] = [];

  for (const slot of team.golfers) {
    const golfer = golferScores.get(slot.golfer_id);

    if (!golfer) {
      continue;
    }

    // DQ or WD after starting -> team is disqualified
    if (isEliminatedStatus(golfer.status)) {
      return {
        team,
        totalScore: null,
        isDisqualified: true,
        tier4Score: null,
        tier2CombinedScore: null,
        rank: null,
      };
    }

    // Use score_to_par (works for both active players and those who missed the cut,
    // since ESPN keeps their score frozen at their 36-hole total)
    const score = golfer.score_to_par ?? 0;
    total += score;

    // Track tiebreaker values
    if (slot.tier === 4) {
      tier4Score = score;
    } else if (slot.tier === 2) {
      tier2Scores.push(score);
    }
  }

  const tier2Combined = tier2Scores.length > 0 ? tier2Scores.reduce((a, b) => a + b, 0) : null;

  return {
    team,
    totalScore: total,
    isDisqualified: false,
    tier4Score,
    tier2CombinedScore: tier2Combined,
    rank: null,
  };
}

/**
 * Check if two scored teams are tied after all tiebreakers.
 */
function isTied(a: ScoredTeam, b: ScoredTeam): boolean {
  return (
    a.totalScore === b.totalScore &&
    a.tier4Score === b.tier4Score &&
    a.tier2CombinedScore === b.tier2CombinedScore
  );
}

/**
 * Score all teams and return them ranked lowest-score-first.
 * Disqualified teams sort to the bottom.
 *
 * Tiebreaker order:
 *   1. Lower total score wins.
 *   2. Lower tier-4 (51+) golfer score.
 *   3. Lower combined tier-2 (11-30) golfer scores.
 *   4. If still tied, they share the position (split prize money).
 */
export function rankTeams(
  teams: TeamForScoring[],
  golferScores: Map<string, GolferScoreData>
): ScoredTeam[] {
  const scored = teams.map((t) => calculateTeamScore(t, golferScores));

  // Sort by: DQ status, total score, tier4 score, tier2 combined score
  scored.sort((a, b) => {
    // DQ'd teams sort to bottom
    if (a.isDisqualified && !b.isDisqualified) return 1;
    if (!a.isDisqualified && b.isDisqualified) return -1;
    if (a.isDisqualified && b.isDisqualified) return 0;

    // Compare total scores (lower is better)
    const aTotal = a.totalScore ?? 999;
    const bTotal = b.totalScore ?? 999;
    if (aTotal !== bTotal) return aTotal - bTotal;

    // Tiebreaker 1: tier-4 score
    const aTier4 = a.tier4Score ?? 999;
    const bTier4 = b.tier4Score ?? 999;
    if (aTier4 !== bTier4) return aTier4 - bTier4;

    // Tiebreaker 2: combined tier-2 score
    const aTier2 = a.tier2CombinedScore ?? 999;
    const bTier2 = b.tier2CombinedScore ?? 999;
    return aTier2 - bTier2;
  });

  // Assign ranks (ties get the same rank; DQ'd teams are unranked)
  let activePosition = 0;
  for (let i = 0; i < scored.length; i++) {
    const st = scored[i];

    if (st.isDisqualified) {
      st.rank = null;
      continue;
    }

    activePosition++;

    if (activePosition === 1) {
      st.rank = 1;
    } else {
      // Find the previous non-DQ team
      let prev: ScoredTeam | null = null;
      for (let j = i - 1; j >= 0; j--) {
        if (!scored[j].isDisqualified) {
          prev = scored[j];
          break;
        }
      }

      if (prev && isTied(prev, st)) {
        st.rank = prev.rank;
      } else {
        st.rank = activePosition;
      }
    }
  }

  return scored;
}

/**
 * Check if any golfer on the team has WD or DQ status.
 */
export function checkTeamDisqualification(
  team: TeamForScoring,
  golferScores: Map<string, GolferScoreData>
): boolean {
  for (const slot of team.golfers) {
    const golfer = golferScores.get(slot.golfer_id);
    if (golfer && isEliminatedStatus(golfer.status)) {
      return true;
    }
  }
  return false;
}
